npm/debugTools/dataplane/tupleProcessor.go

```diff
@@ -0,0 +1,431 @@
+package dataplane
+
+import (
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+
+	"github.com/Azure/azure-container-networking/npm"
+	"github.com/Azure/azure-container-networking/npm/debugTools/pb"
+	"github.com/Azure/azure-container-networking/npm/util"
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+// Tuple struct
+type Tuple struct {
+	RuleType  string `json:"ruleType"`
+	Direction string `json:"direction"`
+	SrcIP     string `json:"srcIP"`
+	SrcPort   string `json:"srcPort"`
+	DstIP     string `json:"dstIP"`
+	DstPort   string `json:"dstPort"`
+	Protocol  string `json:"protocol"`
+}
+
+// Input struct
+type Input struct {
+	Content string
+	Type    InputType
+}
+
+// InputType indicates allowed typle for source and destination input
+type InputType int32
+
+const (
+	// IPADDRS indicates the IP Address input type
+	IPADDRS InputType = 0
+	// PODNAME indicates the podname input type
+	PODNAME InputType = 1
+	// EXTERNAL indicates the external input type
+	EXTERNAL InputType = 2
+)
+
+// GetNetworkTuple read from node's NPM cache and iptables-save and
+// returns a list of hit rules between the source and the destination in
+// JSON format and a list of tuples from those rules.
+func GetNetworkTuple(src, dst *Input) ([][]byte, []*Tuple, error) {
+	c := &Converter{}
+
+	allRules, err := c.GetProtobufRulesFromIptable("filter")
+	if err != nil {
+		return nil, nil, fmt.Errorf("error occurred during get network tuple : %w", err)
+	}
+	return getNetworkTupleCommon(src, dst, c.NPMCache, allRules)
+}
+
+// GetNetworkTupleFile read from NPM cache and iptables-save files and
+// returns a list of hit rules between the source and the destination in
+// JSON format and a list of tuples from those rules.
+func GetNetworkTupleFile(
+	src, dst *Input,
+	npmCacheFile string,
+	iptableSaveFile string,
+) ([][]byte, []*Tuple, error) {
+
+	c := &Converter{}
+	allRules, err := c.GetProtobufRulesFromIptableFile(util.IptablesFilterTable, npmCacheFile, iptableSaveFile)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error occurred during get network tuple : %w", err)
+	}
+
+	return getNetworkTupleCommon(src, dst, c.NPMCache, allRules)
+}
+
+// Common function.
+func getNetworkTupleCommon(
+	src, dst *Input,
+	npmCache *NPMCache,
+	allRules []*pb.RuleResponse,
+) ([][]byte, []*Tuple, error) {
+
+	srcPod, err := getCorrespondPod(src, npmCache)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error occurred during get source pod : %w", err)
+	}
+
+	dstPod, err := getCorrespondPod(dst, npmCache)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error occurred during get destination pod : %w", err)
+	}
+
+	hitRules, err := getHitRules(srcPod, dstPod, allRules, npmCache)
+	if err != nil {
```

Comment:
error message will be duplicated if err is not nil.
In getHitRules, it also returns
```go
	if err != nil {
				return nil, fmt.Errorf("error occurred during evaluating source's set info : %w", err)
			}
```
Can you please walk through them to clean-up duplicated error messages?